ferrischat_webserver/src/auth/reset_password.rs

```diff
@@ -0,0 +1,152 @@
+        .ok_or(WebServerError::RandomGenerationFailure)?;
+
+    // Default email.
+    // TODO HTML rather then plaintext
+    // Also encodes the email to be URL-safe, however some work is needed on it still
+    let default_email = format!(
+        "Hey!\n\nWe see you have requested to reset your password. Click here to do so: https://api.ferris.chat/v0/auth/reset/{}.\n\nIf you did not request this, your account may be compromised.\n\n- FerrisChat Team\nhello@ferris.chat",
+        urlencoding::encode(&*token)
+    );
+
+    // Builds the message with a hardcoded subject and sender full name
+    let message = Message::builder()
+        .from(format!("FerrisChat System <{}>", smtp_username).parse()?)
+        .to(user.email.parse()?)
+        .subject("FerrisChat Password Reset")
+        .body(default_email)?;
+
+    // Open a remote, asynchronous connection to the mail server
+    let mailer = AsyncSmtpTransport::<Tokio1Executor>::starttls_relay(smtp_host.as_str())?
+        .credentials(mail_creds)
+        .build();
+
+    // Send the email
+    mailer.send(message).await?;
+
+    let user_properties = format!("{}||||{}", user.id, hashed_password);
+
+    // writes the token to redis.
+    // The reason we use the token as the key rather then the value is so we can check against it more easily later, when it's part of the URL.
+    redis
+        .set_ex::<String, String, String>(
+            format!("password-reset:{}", token),
+            user_properties,
+            86400,
+        )
+        .await?;
+
+    Ok(crate::Json::new(
+        SuccessJson::new("Sent password reset, please check your email.".to_string()),
+        200,
+    ))
+}
+
+/// GET /v0/auth/reset/{token}
+/// Verifies the user's email when they click the link mailed to them.
+pub async fn verify_password_reset(
+    Path(token): Path<String>,
+) -> Result<crate::Json<SuccessJson>, WebServerError> {
+    let db = get_db_or_fail!();
+
+    let redis_key = format!("password-reset:{}", token);
+    let id_and_hashed_password_as_string = ferrischat_redis::redis::cmd("GETDEL")
+        .arg(redis_key)
+        .query_async::<_, Option<String>>(
+            &mut REDIS_MANAGER
+                .get()
+                .ok_or(WebServerError::MissingRedis)?
+                .get()
+                .await?,
+        )
+        .await?
+        .ok_or_else(|| {
+            ErrorJson::new_404("This token has expired or was not found.".to_string())
+        })?;
+    let id_and_hashed_password_as_vec: Vec<&str> =
+        id_and_hashed_password_as_string.split("||||").collect();
+    // Tell the database to set their verified field to true! The user is now verified.
+    sqlx::query!(
+        "UPDATE users SET password = $1 WHERE id = $2",
+        id_and_hashed_password_as_vec[0],
+        u128_to_bigdecimal!(id_and_hashed_password_as_vec[1]
+            .to_string()
+            .parse::<u128>()
+            .unwrap()),
```

Comment:
I'm gonna be nitpicky here and say to handle this error: the only way it could error is due to human error at this point, which is definitely possible.
```suggestion
            .map_err(|e| ErrorJson::new_500(format!("failed to parse user ID: {}", e), false, None)))?,
```